### PR TITLE
Implement SwayClientUpdate class

### PR DESCRIPTION
**Pull request by xezon**
Does not affect all trees. On Seaside Mutiny (2) map there is this tree that uses this module.

shot_20220122_202015_2

### CODE REVIEW

File: src/game/client/draw/drawable.cpp [hunk 1]
```diff
@@ -26,6 +28,18 @@ const Matrix3D *Drawable::Get_Transform_Matrix() const
     return Thing::Get_Transform_Matrix();
 }
 
+bool Drawable::Is_Visible() const
+{
+    const DrawModule **draw_modules = Get_Draw_Modules();
+
+    for (const DrawModule *draw_module = draw_modules[0]; draw_modules != nullptr; ++draw_module) {
+        if (draw_module->Is_Visible()) {
```
Comment (xezon) [inline src/game/client/draw/drawable.cpp:36]: I did not see the vtable mappings of `DrawModule` class, but guessed `Is_Visible` looks like the right choice.

File: src/game/common/namekeygenerator.cpp [hunk 2]
```diff
@@ -155,3 +155,8 @@ void NameKeyGenerator::Free_Sockets()
         m_sockets[i] = nullptr;
     }
 }
+
+NameKeyType Name_To_Key(const char *name)
```
Comment (xezon) [inline src/game/common/namekeygenerator.cpp:159]: As seen in WorldBuilder.

File: src/game/logic/object/update/swayclientupdate.cpp [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+void SwayClientUpdate::Load_Post_Process()
+{
+    ClientUpdateModule::Load_Post_Process();
+    Update_Sway();
+}
+
+NameKeyType SwayClientUpdate::Get_Module_Name_Key() const
+{
+    if (s_swayClientUpdateKey == NAMEKEY_INVALID) {
```
Comment (xezon) [inline src/game/logic/object/update/swayclientupdate.cpp]: In original it tests a static bool here, but I thought we can simplify this and simply check for NAMEKEY_INVALID.

File: src/game/logic/object/update/swayclientupdate.cpp [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+SwayClientUpdate *SwayClientUpdate::Friend_New_Module_Instance(Thing *thing, const ModuleData *module_data)
+{
+    return new SwayClientUpdate(thing, module_data);
+}
+
+void SwayClientUpdate::CRC_Snapshot(Xfer *xfer)
+{
+    DrawableModule::CRC_Snapshot(xfer);
```
Comment (tomsons26) [inline src/game/logic/object/update/swayclientupdate.cpp]: ClientUpdateModule::CRC_Snapshot(xfer);

File: src/game/logic/object/update/swayclientupdate.cpp (hunk 3 above)
Comment (tomsons26) [reply src/game/logic/object/update/swayclientupdate.cpp]: its compiler generated on statics
https://github.com/TheAssemblyArmada/Thyme/blob/develop/src/game/logic/object/update/laserupdate.cpp#L78

Comment (xezon) [reply src/game/logic/object/update/swayclientupdate.cpp]: Fixed.

File: src/game/logic/object/update/swayclientupdate.cpp (hunk 4 above)
Comment (xezon) [reply src/game/logic/object/update/swayclientupdate.cpp]: Fixed.

File: src/game/logic/object/update/swayclientupdate.cpp [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+            m_curValue -= GAMEMATH_PI2;
+        }
+        const float angle = GameMath::Cos(m_curValue);
+        const float new_angle = angle * m_curAngleLimit + m_leanAngle;
+        const float delta_angle = new_angle - m_curAngle;
+        const float rotx = -delta_angle * breeze->sway_direction.x;
+        const float roty = delta_angle * breeze->sway_direction.y;
+
+        Matrix3D new_matrix = drawable->Get_Instance_Matrix();
+        new_matrix.Rotate_X(rotx);
```
Comment (tomsons26) [inline src/game/logic/object/update/swayclientupdate.cpp]: In_Place_Pre_Rotate_X
In_Place_Pre_Rotate_Y

File: src/game/logic/object/update/swayclientupdate.cpp [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    return s_swayClientUpdateKey;
+}
+
+void SwayClientUpdate::Client_Update()
+{
+    if (m_swaying) {
+        Drawable *drawable = Get_Drawable();
+        const BreezeInfo *breeze = g_theScriptEngine->Get_Breeze_Info();
+
+        if (m_curVersion == breeze->version) {
```
Comment (tomsons26) [inline src/game/logic/object/update/swayclientupdate.cpp]: flip this for consistency with assembly

File: src/game/logic/object/update/swayclientupdate.cpp (hunk 5 above)
Comment (xezon) [reply src/game/logic/object/update/swayclientupdate.cpp]: Tested. Visually looks the same. Trees sway the same way with same speeds. Was this supposed to change something?

File: src/game/logic/object/update/swayclientupdate.cpp (hunk 6 above)
Comment (xezon) [reply src/game/logic/object/update/swayclientupdate.cpp]: What does flip mean?

File: src/hooker/setuphooks_zh.cpp [hunk 7]
```diff
@@ -1997,4 +1998,13 @@ void Setup_Hooks()
     Hook_Method(0x0072C180, &BezierSegment::Get_Segment_Points);
     Hook_Method(0x0072C2A0, &BezierSegment::Get_Approximate_Length);
     Hook_Method(0x0072C440, &BezierSegment::Split_Segment_At_T);
+
+    // swayclientupdate.h
+    Hook_Method(0x006885A0, &SwayClientUpdate::Hook_Ctor);
+    Hook_Any(0x00688660, SwayClientUpdate::~SwayClientUpdate);
```
Comment (tomsons26) [inline src/hooker/setuphooks_zh.cpp]: can't hook the dtor as its inlined in the msvc made vector destructor at 0x00688660

File: src/game/logic/object/update/swayclientupdate.cpp (hunk 6 above)
Comment (xezon) [reply src/game/logic/object/update/swayclientupdate.cpp]: Done.

File: src/hooker/setuphooks_zh.cpp (hunk 7 above)
Comment (xezon) [reply src/hooker/setuphooks_zh.cpp]: Removed.

File: src/game/client/draw/drawable.cpp (hunk 1 above)
Comment (tomsons26) [reply src/game/client/draw/drawable.cpp:36]: yeah correct

File: src/game/common/namekeygenerator.cpp (hunk 2 above)
Comment (tomsons26) [reply src/game/common/namekeygenerator.cpp:159]: make it a inline

File: src/game/logic/object/update/swayclientupdate.cpp (hunk 5 above)
Comment (tomsons26) [reply src/game/logic/object/update/swayclientupdate.cpp]: its possible it won't in place pre and pre sometimes result in same results as normal rotate, but not always, don't know much of matrix maths to tell you how that goes
In_Place_Pre is def the right functions tho

Comment (xezon) [reply src/game/logic/object/update/swayclientupdate.cpp]: Fixed.

File: src/game/common/namekeygenerator.cpp (hunk 2 above)
Comment (xezon) [reply src/game/common/namekeygenerator.cpp:159]: It's ok as is. Compiler will compile out small functions regardless of keyword.

**Review by tomsons26 — APPROVED** [review]

**Pull request merged**